src/ray/raylet/worker_pool.cc

```diff
@@ -41,9 +44,11 @@ namespace raylet {
 /// A constructor that initializes a worker pool with
 /// (num_worker_processes * num_workers_per_process) workers for each language.
 WorkerPool::WorkerPool(
-    int num_worker_processes, int num_workers_per_process, int num_cpus,
+    int num_worker_processes, int num_workers_per_process,
+    int maximum_startup_concurrency,
     const std::unordered_map<Language, std::vector<std::string>> &worker_commands)
-    : num_workers_per_process_(num_workers_per_process), num_cpus_(num_cpus) {
+    : num_workers_per_process_(num_workers_per_process),
+      maximum_startup_concurrency_(maximum_startup_concurrency) {
```

Comment:
we should set `maximum_startup_concurrency_` to at least 1, in case user pass 0 or a negative number by accident.

Reply:
Hm, I think 0 should be allowed as a way to specify that no workers should be created.

Reply:
If `maximum_startup_concurrency_` is 0, no workers will ever be created. I think that's useless? 

Did you mean the number of initial workers to create when starting? that is another parameter, `num_worker_processes`.

Reply:
So, it's not necessarily useless, because you can still submit tasks and get objects. It just means that that node won't execute any tasks.

However, I think the appropriate way to specify that no tasks should be execute on a node is to set the resources to be 0 instead of setting the `maximum_startup_concurrency_` to 0. E.g., if we have 1 CPU resource and `maximum_startup_concurrency_` then that could lead to hanging.

Maybe we should simply do `RAY_CHECK(maximum_startup_concurrency_ > 0)`.

Reply:
1) I'm afraid I still don't understand what the benefit is to add a no-op node to the cluster. One thing I can think of is that it allows users to add their laptops to a cluster and run drivers from the laptops without logging onto a prod machine. But this problem should be solved by job submission API. Am I missing something?

2) I'm not sure if `RAY_CHECK` is too strong here. An alternative way is setting the value to 1 and printing a warning. But either way is fine.

Reply:
We should keep the ability to create a ray cluster node that's guaranteed to accept no work. I commonly use this for testing local vs. remote task execution throughput and object store throughput.

Reply:
@atumanov thanks, that's a good point.

Reply:
@atumanov @raulchen I think the "correct" way to specify that no tasks should run at a node is to specify 0 resources as opposed to limiting the worker startup concurrency.

Especially since a 0-resource node can still execute tasks that require 0 resources, so we may need to start workers even on nodes with 0 resources.

@raulchen about the `RAY_CHECK`, changing the value sounds like it could potentially hide errors or cause confusion. If `0` is not a valid input, then I'd prefer to fail immediately (of course it should not be possible to trigger the fatal code path from the Python/Java side).

Reply:
@robertnishihara , yeah that's right. I think of ray nodes in terms of their multi-dimensional resource capacity. I think that the only units of compute that may require zero resources are actor tasks, in which case the actor creation task must've required resources, which would not have been placed on the zero-cpu node. It's a good sanity check and probably a good time to ensure that we don't have (don't support?) cases where the sum of actor creation resreq and actor task resreq is ever zero. That just wouldn't make sense to me.

FWIW, thus far, whenever I wanted a "no-op" ray node (one that accepts no work), I  specify `--num-cpus 0 --num-workers 0`.

Reply:
Agreed that specifying 0 resource is the proper way for no-op nodes.

@robertnishihara regarding the `RAY_CHECK`, your reasoning makes sense to me. Previously I was concerned that users might accidentally specify 0 cpu and have no idea what went wrong (new users may not know where to find the log file). Now since we ensure the argument is at least 1 in Python & Java. This won't happen.